### PR TITLE
Add new test cases for gateway server

### PR DESCRIPTION
**Pull request by ModiCodeCraftsman**
This PR adds comprehensive unit tests and supporting test utilities for the gateway server component. The main changes include:

New Tests:
- Added gateway_req_body_test.go to cover request body handling logic, including routing, error handling, and edge cases.
- Added gateway_test_helpers.go with mock implementations for easier and more robust testing.
- Extended gateway_test.go with new tests for pod selection and routing strategies.

Test Infrastructure Improvements:

- Refactored and improved the e2e test script (run-e2e-tests.sh) for better port-forward management and cleanup.

[Issue](https://github.com/vllm-project/aibrix/issues/1216)

### CODE REVIEW

File: pkg/plugins/gateway/algorithms/router.go [hunk 1]
```diff
@@ -31,6 +34,8 @@ var routerConstructor = map[types.RoutingAlgorithm]types.RouterProviderRegistrat
 
 // Validate validates if user provided routing routers is supported by gateway
 func Validate(algorithms string) (types.RoutingAlgorithm, bool) {
+	routerMu.RLock()
```
Comment (varungup90) [inline pkg/plugins/gateway/algorithms/router.go:37]: due to data race condition for router factory variable?

File: pkg/plugins/gateway/gateway_test.go [hunk 2]
```diff
@@ -143,3 +149,146 @@ func Test_buildEnvoyProxyHeaders(t *testing.T) {
...
+	// Define test cases for different pod selection and error scenarios
+	tests := []struct {
+		name          string
+		pods          types.PodList
+		mockSetup     func(*mockRouter, types.RoutingAlgorithm)
+		expectedError bool
+		expectedPodIP string
+	}{
+		{
+			name: "routing.Select returns error",
```
Comment (varungup90) [inline pkg/plugins/gateway/gateway_test.go]: In this test, router.Route() is returning error, not the routing.Select(). Please update the test description.

File: pkg/plugins/gateway/gateway_test.go [hunk 3]
```diff
@@ -143,3 +149,146 @@ func Test_buildEnvoyProxyHeaders(t *testing.T) {
...
+			pods: &mockPodList{pods: []*v1.Pod{{
```
Comment (varungup90) [inline pkg/plugins/gateway/gateway_test.go]: Can you help me understand why mockPodList is needed here, directly using podList would be fine too.

File: pkg/plugins/gateway/gateway_test.go [hunk 4]
```diff
@@ -143,3 +149,146 @@ func Test_buildEnvoyProxyHeaders(t *testing.T) {
...
+				// No expectations needed as no ready pods
+			},
+			expectedError: true,
+		},
+		{
+			name: "single ready pod",
+			pods: &mockPodList{pods: []*v1.Pod{{
+				Status: v1.PodStatus{
+					PodIP:      "1.2.3.4",
+					Conditions: []v1.PodCondition{{Type: v1.PodReady, Status: v1.ConditionTrue}},
```
Comment (varungup90) [inline pkg/plugins/gateway/gateway_test.go:224]: Also add one more pod with ConditionFalse.

File: pkg/plugins/gateway/gateway_test.go (hunk 3 above)
Comment (varungup90) [reply pkg/plugins/gateway/gateway_test.go]: For these tests, even random router can be used.

File: pkg/plugins/gateway/algorithms/router.go (hunk 1 above)
Comment (ModiCodeCraftsman) [reply pkg/plugins/gateway/algorithms/router.go:37]: Yes, this ensures thread-safe concurrent R/W and prevents data race conditions.

File: pkg/plugins/gateway/gateway_test.go (hunk 3 above)
Comment (ModiCodeCraftsman) [reply pkg/plugins/gateway/gateway_test.go]: > Can you help me understand why mockPodList is needed here, directly using podList would be fine too.

The original motivation for using a mockedPodList type was to avoid potential issues arising from the internal logic of certain PodArray methods (such as ListByIndex and related indexing behavior) during testing. This allowed us to isolate test scenarios from the implementation details of PodArray. However, following the review and for consistency, the mock type has now been removed and all relevant tests have been updated to use the actual PodArray type directly.

> For these tests, even random router can be used.

My intention with the current approach was to use a mock router to precisely control the behavior of the Route function, including simulating errors and verifying whether Route is called in specific scenarios. This allows for more granular testing of error handling and call expectations within the server logic.

If your suggestion was to simplify the tests by using the actual random router implementation, could you please clarify if there are specific scenarios where you think the mock is unnecessary, or if there’s a particular benefit you see in using the real router here? I want to make sure I fully understand your point and am happy to adjust the tests if needed.

File: pkg/plugins/gateway/gateway_test.go (hunk 4 above)
Comment (ModiCodeCraftsman) [reply pkg/plugins/gateway/gateway_test.go:224]: Added the "single ready pod out of two" test case

File: pkg/plugins/gateway/gateway_test.go (hunk 2 above)
Comment (ModiCodeCraftsman) [reply pkg/plugins/gateway/gateway_test.go]: Done.

File: pkg/plugins/gateway/gateway_req_body_test.go [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+						assert.Contains(t, []string{"1.2.3.4:8000", "5.6.7.8:8000"}, targetPodIP, "Target pod IP should be one of the pod IPs from the mock setup")
+					}
+				}
+				assert.True(t, foundRoutingStrategy, "HeaderRoutingStrategy not found")
+				assert.True(t, foundTargetPod, "HeaderTargetPod not found")
+			},
+			checkStream: false,
+		},
+		{
+			name:        "no routable pods available - should return ServiceUnavailable",
```
Comment (varungup90) [inline pkg/plugins/gateway/gateway_req_body_test.go:338]: There are 4 tests here.
- pod len == 0 (missing).
- pod condition false (present)
- pod IP not set (present)
- pod in terminating state (missing).

Can you add the missing two cases.

File: pkg/plugins/gateway/gateway_test.go (hunk 3 above)
Comment (varungup90) [reply pkg/plugins/gateway/gateway_test.go]: I think its ok to use mock router for most gateway tests. Later I can add one simple test e2e without mock router and mock cache.

**Comment by varungup90** [comment]
lgtm. Nit comment to add two missing cases for service unavailable.

File: pkg/plugins/gateway/gateway_req_body_test.go (hunk 5 above)
Comment (ModiCodeCraftsman) [reply pkg/plugins/gateway/gateway_req_body_test.go:338]: Sure, done.

**Comment by ModiCodeCraftsman** [comment]
@varungup90 Could you please review and resolve the open discussions if you’re satisfied with the changes? I believe it’s best for the reviewer to close discussions once their requests have been addressed, to ensure nothing is missed. Let me know if there’s anything else you’d like me to update. Thanks!

**Comment by Jeffwan** [comment]
@varungup90 any further comments for this change?

**Review by varungup90 — APPROVED** [review]
